File: Java/src/com/loggersoft/performance/DcfFile.java
package com.loggersoft.performance;

import java.io.*;

public class DcfFile {
    private File file;
    private static final String CLUST_ID = "ClustID";

    public interface IEventHandler {

        boolean segmentRetrieved(DcfSegment segment);

        void invalidLine(String line);
    }

    public DcfFile(File file) {
        this.file = file;
    }

    public void processFile(IEventHandler handler) throws IOException {
        try(BufferedReader in = new BufferedReader(new InputStreamReader(new FileInputStream(file), "UTF-8"))) {
            String line;
            while ((line = in.readLine()) != null) {
                final String token = line.trim();
                if (!token.startsWith("SEG:")) {
                    continue;
                }
                final String parts[] = line.split(":");
                if (parts.length < 6) {
                    handler.invalidLine(line);
                    continue;
                }
                int clustIdIndex = -1;
                for (int index = 5; index < parts.length; index++) {
                    if (parts[index].startsWith(CLUST_ID)) {
                        clustIdIndex = index;
                        break;
                    }
                }
                if (clustIdIndex < 0 || clustIdIndex + 1 >= parts.length) {
                    handler.invalidLine(line);
                    continue;
                }
                try {
                    if (!handler.segmentRetrieved(new DcfSegment(new Position(Double.parseDouble(parts[2]), Double.parseDouble(parts[1])),
                                                                 new Position(Double.parseDouble(parts[4]), Double.parseDouble(parts[3])),
                                                                 parts[clustIdIndex + 1]))) {
                        break;
                    }
                } catch (NumberFormatException e) {
                    handler.invalidLine(line);
                    continue;
                }
            }
        }
    }
}

File: Kotlin/src/com/loggersoft/kotlin/performance/DcfFile.kt
package com.loggersoft.kotlin.performance

import java.io.BufferedReader
import java.io.File
import java.io.FileInputStream
import java.io.InputStreamReader

class DcfFile(private val file: File) {

    fun processFile(segmentRetrieved: (DcfSegment) -> Boolean, invalidLine: (String) -> Unit) {
        BufferedReader(InputStreamReader(FileInputStream(file), "UTF-8")).use {
            while (true) {
                val line = it.readLine() ?: break
                val token = line.trim()
                if (!token.startsWith("SEG:")) continue
                val parts = token.split(':')
                if (parts.size < 6) {
                    invalidLine(line)
                    continue
                }
                val clustIdIndex = (5 until parts.size).firstOrNull { parts[it].startsWith("ClustID") }
                        ?: -1
                if (clustIdIndex < 0 || clustIdIndex + 1 >= parts.size) {
                    invalidLine(line)
                    continue
                }
                try {
                    if (!segmentRetrieved(DcfSegment(Position(parts[2].toDouble(), parts[1].toDouble()),
                                                     Position(parts[4].toDouble(), parts[3].toDouble()),
                                                     parts[clustIdIndex + 1]))) {
                        break
                    }
                } catch (e: NumberFormatException) {
                    invalidLine(line)
                }
            }
        }
    }
}